src/rtc_pl031.rs

```diff
@@ -0,0 +1,612 @@
+        data = 123u32.to_le_bytes();
+        match rtc.write(RTCMR, &data) {
+            Err(Error::BadWriteOffset(_)) => (),
+            _ => unreachable!(),
+        }
+
+        // Reading the match register should do nothing.
+        rtc.read(RTCMR, &mut data);
+        let v = u32::from_le_bytes(data);
+        assert_eq!(v, 123);
+    }
+
+    #[test]
+    fn test_load_register() {
+        // Read and write to the Load Register.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data = [0; 4];
+
+        // Get the RTC value with a counter_offset of 0 (the initial value).
+        rtc.read(RTCDR, &mut data);
+        let old_val = u32::from_le_bytes(data);
+
+        // Write system time since UNIX_EPOCH in seconds to the Load Register
+        let counter_offset = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+        data = (counter_offset as u32).to_le_bytes();
+        rtc.write(RTCLR, &data).unwrap();
+
+        // Read the counter_offset from the Load Register and verify it
+        // matches the value we just loaded.
+        rtc.read(RTCLR, &mut data);
+        assert_eq!((counter_offset as u32), u32::from_le_bytes(data));
+
+        // Confirm that the new RTC value is greater than the old
+        rtc.read(RTCDR, &mut data);
+        let new_val = u32::from_le_bytes(data);
+        assert!(new_val > old_val);
+    }
+
+    #[test]
+    fn test_interrupt_mask_set_clear_register() {
+        // Read and write to IMSC register.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data: [u8; 4];
+
+        // Set the mask bit and confirm the interrupt event is triggered.
+        data = 1u32.to_le_bytes();
+        rtc.write(RTCIMSC, &data).unwrap();
+        // The interrupt line should be on.
+        assert!(rtc.interrupt_evt.read().unwrap() == 1);
+
+        // Read the mask bit and confirm it's set.
+        rtc.read(RTCIMSC, &mut data);
+        let v = u32::from_le_bytes(data);
+        assert_eq!(1, v);
+
+        // Clear the mask bit and confirm it's cleared.
+        data = 0u32.to_le_bytes();
+        rtc.write(RTCIMSC, &data).unwrap();
+        rtc.read(RTCIMSC, &mut data);
+        let v = u32::from_le_bytes(data);
+        assert_eq!(0, v);
+    }
+
+    #[test]
+    fn test_interrupt_clear_register() {
+        // Read and write to the ICR register.
+        let intr_evt = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+        let mut rtc = RTC::new(intr_evt);
+        let mut data: [u8; 4];
+
+        // Write to the interrupt clear register and confirm the interrupt
+        // event is triggered.
+        data = 1u32.to_le_bytes();
+        rtc.write(RTCICR, &data).unwrap();
+        // The interrupt line should be on.
+        assert!(rtc.interrupt_evt.read().unwrap() == 1);
```

Comment:
This test seems backwards to me. The PL031 manual says the following about `RTCICR`:

> Writing 1 to bit[0] clears the interrupt.

I would therefore expect that the interrupt line should *not* be on. This is in the original Firecracker code already, though.

Reply:
We fixed this here https://github.com/firecracker-microvm/firecracker/pull/2453/commits/f8abd9c01763f1815f436b4426ba806c403805ed.

Reply:
I was confused here and assumed [something] needed to be made aware that the interrupt had been cleared.  I didn't (still don't) have the clearest understanding of how interrupts interact between the hypervisor, vmm, and vm.  Anyway...

I've now removed all uses of `interrupt_evt.trigger()`, since nothing in my current implementation needs it.  This test now writes to `RTCICR` and then confirms that the raw and masked interrupts are indeed clear.

I've also tried to mirror some of the updates to firecracker, as posted by @dianpopa.